docs/design/2022-03-31-concurrent-ddl-framawork.md

```diff
@@ -0,0 +1,233 @@
+| job_meta      | blob       | YES  |      |
++---------------+------------+------+------+
+```
+
+In bootstrap step, TiDB will build these tables meta and put it into tikv directly. For new cluster, TiDB will also build `mysql` schema meta.
+
+### DDL operations
+
+Use the following DDL operations to manage the DDL jobs:
+```sql
+insert into mysql.tidb_ddl_job values (...) -- add ddl jobs
+delete from mysql.tidb_ddl_job where job_id = ... -- delete ddl jobs
+update mysql.tidb_ddl_job set job_meta = ... where job_id = ... -- update ddl jobs
+select * from mysql.tidb_ddl_job -- get ddl jobs
+```
+
+### DDL job manager
+
+DDL job manager will find the runnable DDL job and dispatch the jobs to DDL workers.
+```golang
+for {
+    if !isOwner {
+        // sleep a while
+        continue
+    }
+
+    // if there are job coming
+    select {
+        case <-ddlJobCh:
+        case <-ticker.C:
+        case <-notifyDDLJobByEtcdCh:
+    }
+
+    job = findrunnableJob()
+    if job != nil {
+        waitForFreeWorker();
+        go runJob(job)
+    }
+}
+```
+
+To prevent all the workers in worker pool occupied by the long run job like `add index`, TiDB will divide the pool into two types: `addIndex` worker pool and `general` worker pool, then the DDL job manager will be
+
+```golang
+for {
+    // ...
+    generalJob = findRunnableGeneralJob()
+    reorgJob = findRunnableReorgJob()
+    if generalJob != nil {
+        waitForFreeGeneralWorker();
+        go runGeneralJob(job)
+    }
+    if reorgJob != nil {
+        waitForFreeReorgWorker();
+        go runReorgJob(job)
+    }
+}
+```
+
+The runnable job defines as
+1. not running.
+2. the minimum job id on the same table.
+3. no smaller `drop schema` job id on the same schema.
+4. no smaller job id on the same schema if the job type is `drop schema`.
+
+to get a general job, we can use SQL
+```sql
+select * from mysql.tidb_ddl_job where not reorg;
+```
+and return the first record.
+
+Now we will describe how to achieve the 4 rules above:
+
+Rule 1 and Rule 2: maintain a running job set, group by the table id and find the minimum one, and SQL can be written to
+```sql
+select * from mysql.tidb_ddl_job where job_id in (select min(job_id) from mysql.tidb_ddl_job where not reorg and job_id not in ({running job id}) group by table_id);
+```
+
+Then, we find some jobs and check Rule 3 and Rule 4:
+
+Rule 3: check if there is a smaller `drop schema` job id on the same schema.
+```sql
+select * from mysql.tidb_ddl_job where is_drop_schema and job_id < {job.id} limit 1;
+```
+Rule 4: if the job is `drop schema`, check if there is a smaller job id on the same schema.
+```sql
+select * from mysql.tidb_ddl_job where job_id < {job.id} limit 1;
+```
+
+For reorg job, the process is almost the same.
+
+![ddl-job-manager](./imgs/ddl-job-manager.png)
+
+### Show DDL jobs
+
+get DDL jobs from the tables:
+```sql
+select * from mysql.tidb_ddl_job;
+select * from mysql.tidb_history_job;
```

Comment:
Compatible with the original code, do you want to add `limit 10` here?

Reply:
yes, the original code has a limit 10, here just means we can use these SQLs to get the ddl jobs. We will add a `limit 10` in the implementation